File: windows/MISILNative.Core/Agerbot/AgerbotModelUpdateController.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using MISILNative.Core.Distribution;

namespace MISILNative.Core.Agerbot;

public sealed class AgerbotModelUpdateController : INotifyPropertyChanged, IDisposable
{
    private readonly string _managedRoot;
    private readonly AgerbotModelReleaseService _releases;
    private readonly AgerbotModelDownloadService _downloads;
    private readonly AgerbotModelInstallationManager _installation;
    private readonly IAgerbotCandidateValidator _validator;
    private readonly AgerbotSettingsStore _settings;
    private readonly AgerbotCurrentModelStore _currentStore;
    private readonly AgerbotFailedVersionStore _failedStore;
    private readonly AgerbotModelActivationService _activation;
    private readonly IAgerbotModelRuntimeActivator _runtimeActivator;
    private readonly SemaphoreSlim _operation = new(1, 1);
    private CancellationTokenSource? _cancellation;
    private AgerbotRemoteModelCandidate? _availableCandidate;
    private AgerbotModelUpdateState _state = new(AgerbotModelUpdateStatus.Idle, "Sin comprobar");

    public AgerbotModelUpdateController(
        string managedRoot,
        AgerbotModelReleaseService releases,
        AgerbotModelDownloadService downloads,
        AgerbotModelInstallationManager installation,
        IAgerbotCandidateValidator validator,
        AgerbotSettingsStore settings,
        AgerbotCurrentModelStore currentStore,
        AgerbotFailedVersionStore failedStore,
        AgerbotModelActivationService activation,
        IAgerbotModelRuntimeActivator runtimeActivator)
    {
        _managedRoot = Path.GetFullPath(managedRoot);
        _releases = releases;
        _downloads = downloads;
        _installation = installation;
        _validator = validator;
        _settings = settings;
        _currentStore = currentStore;
        _failedStore = failedStore;
        _activation = activation;
        _runtimeActivator = runtimeActivator;
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    public AgerbotModelUpdateState State
    {
        get => _state;
        private set { _state = value; OnPropertyChanged(); OnPropertyChanged(nameof(IsBusy)); }
    }

    public bool IsBusy => State.Status is AgerbotModelUpdateStatus.Checking
        or AgerbotModelUpdateStatus.Downloading
        or AgerbotModelUpdateStatus.Verifying
        or AgerbotModelUpdateStatus.WaitingForConversation
        or AgerbotModelUpdateStatus.Activating
        or AgerbotModelUpdateStatus.RollingBack;

    public bool HasAvailableUpdate => _availableCandidate != null;

    public async Task CheckForUpdatesAsync(
        bool force = true,
        bool installAutomatically = false,
        ulong diskAvailableBytes = ulong.MaxValue,
        CancellationToken cancellationToken = default)
    {
        if (!await _operation.WaitAsync(0, cancellationToken)) return;
        _cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        try
        {
            _availableCandidate = null;
            OnPropertyChanged(nameof(HasAvailableUpdate));
            if (!force && _settings.Settings.LastUpdateCheckAt is DateTimeOffset last
                && DateTimeOffset.UtcNow - last < TimeSpan.FromHours(6))
            {
                State = new(AgerbotModelUpdateStatus.Idle, "Comprobación reciente");
                return;
            }
            if (string.IsNullOrWhiteSpace(_settings.Settings.RuntimeVersion))
                throw new InvalidOperationException("Instala primero el runtime de Agerbot.");
            State = new(AgerbotModelUpdateStatus.Checking, "Buscando un modelo estable compatible…");
            _settings.Update(value => value.LastUpdateCheckAt = DateTimeOffset.UtcNow);
            var candidate = await _releases.LatestCompatibleAsync(
                _settings.Settings.RuntimeVersion,
                _settings.Settings.ActiveModelVersion,
                _settings.Settings.AllowPrereleaseModels,
                _cancellation.Token);
            string version = candidate.Manifest.Release.Version;
            if (await _failedStore.ContainsAsync(version, _cancellation.Token))
                throw new InvalidOperationException($"Agerbot {version} falló anteriormente y no se reinstalará automáticamente.");
            _availableCandidate = candidate;
            OnPropertyChanged(nameof(HasAvailableUpdate));
            State = new(AgerbotModelUpdateStatus.Available, $"Modelo Agerbot {version} disponible", version, candidate.Manifest.Artifact.SizeBytes);
            if (installAutomatically && _settings.Settings.AutomaticModelUpdates)
                await InstallAvailableUnsafeAsync(_cancellation.Token, diskAvailableBytes);
        }
        catch (OperationCanceledException)
        {
            State = new(AgerbotModelUpdateStatus.Cancelled, "Actualización cancelada");
        }
        catch (NoUpdateAvailableException exception)
        {
            State = new(AgerbotModelUpdateStatus.Idle, exception.Message);
        }
        catch (Exception exception)
        {
            State = new(AgerbotModelUpdateStatus.Failed, exception.Message);
        }
        finally { EndOperation(); }
    }

    public async Task InstallAvailableAsync(ulong diskAvailableBytes, CancellationToken cancellationToken = default)
    {
        if (_availableCandidate == null || !await _operation.WaitAsync(0, cancellationToken)) return;
        _cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        try { await InstallAvailableUnsafeAsync(_cancellation.Token, diskAvailableBytes); }
        catch (OperationCanceledException)
        {
            State = new(AgerbotModelUpdateStatus.Cancelled, "Descarga cancelada; MISIL podrá reanudarla.");
        }
        catch (Exception exception)
        {
            State = new(AgerbotModelUpdateStatus.Failed, exception.Message);
        }
        finally { EndOperation(); }
    }

    public async Task RollbackAsync(CancellationToken cancellationToken = default)
    {
        string? previousVersion = _settings.Settings.PreviousModelVersion;
        if (string.IsNullOrWhiteSpace(previousVersion) || !await _operation.WaitAsync(0, cancellationToken)) return;
        _cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        try
        {
            State = new(AgerbotModelUpdateStatus.RollingBack, $"Volviendo a Agerbot {previousVersion}", previousVersion);
            string runtimeVersion = _settings.Settings.RuntimeVersion
                ?? throw new InvalidOperationException("No hay un runtime instalado.");
            var discovery = new AgerbotModelDiscoveryService(runtimeVersion);
            var models = await discovery.DiscoverAsync(_managedRoot, cancellationToken: _cancellation.Token);
            var model = models.FirstOrDefault(item => item.Manifest.Model.Version == previousVersion)
                ?? throw new InvalidOperationException("La versión anterior ya no está disponible localmente.");
            var current = await CurrentRecordAsync(_cancellation.Token);
            var target = new AgerbotCurrentModelRecord
            {
                ActiveVersion = previousVersion,
                PreviousVersion = current?.ActiveVersion,
                ActivatedAt = DateTimeOffset.UtcNow,
                CheckpointPath = model.CheckpointPath,
                ManifestPath = model.ManifestPath
            };
            bool success = await _activation.ActivateAsync(target, current, _runtimeActivator, _cancellation.Token);
            State = success
                ? new(AgerbotModelUpdateStatus.Installed, $"Agerbot {previousVersion} está activo", previousVersion, Progress: 1)
                : new(AgerbotModelUpdateStatus.Failed, "No se pudo volver al modelo anterior.");
        }
        catch (OperationCanceledException) { State = new(AgerbotModelUpdateStatus.Cancelled, "Rollback cancelado"); }
        catch (Exception exception) { State = new(AgerbotModelUpdateStatus.Failed, exception.Message); }
        finally { EndOperation(); }
    }

    public async Task RepairAsync(CancellationToken cancellationToken = default)
    {
        if (!await _operation.WaitAsync(0, cancellationToken)) return;
        _cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        try
        {
            string runtimeVersion = _settings.Settings.RuntimeVersion
                ?? throw new InvalidOperationException("No hay un runtime instalado.");
            State = new(AgerbotModelUpdateStatus.Verifying, "Verificando la instalación administrada…");
            var discovery = new AgerbotModelDiscoveryService(runtimeVersion);
            var models = await discovery.DiscoverAsync(_managedRoot, cancellationToken: _cancellation.Token);
            var model = models.FirstOrDefault(item => item.Manifest.Model.Version == _settings.Settings.ActiveModelVersion)
                ?? discovery.Select(models, _settings.Settings.PinnedModelVersion, true, _settings.Settings.AllowPrereleaseModels)
                ?? throw new InvalidOperationException("No hay un modelo local válido para reparar.");
            bool valid = await ValidateWithCpuFallbackAsync(
                model.CheckpointPath,
                model.Manifest.Model.Version,
                _cancellation.Token);
            if (!valid) throw new InvalidOperationException("La instalación local no superó la validación aislada.");
            var previous = await CurrentRecordAsync(_cancellation.Token);
            var record = new AgerbotCurrentModelRecord
            {
                ActiveVersion = model.Manifest.Model.Version,
                PreviousVersion = previous?.ActiveVersion == model.Manifest.Model.Version
                    ? previous.PreviousVersion
                    : previous?.ActiveVersion,
                ActivatedAt = DateTimeOffset.UtcNow,
                CheckpointPath = model.CheckpointPath,
                ManifestPath = model.ManifestPath
            };
            State = new(AgerbotModelUpdateStatus.Activating, "Reiniciando el runtime verificado", record.ActiveVersion);
            if (!await _activation.ActivateAsync(record, previous, _runtimeActivator, _cancellation.Token))
                throw new InvalidOperationException("La reparación falló; MISIL restauró el estado anterior.");
            State = new(AgerbotModelUpdateStatus.Installed, "Instalación de Agerbot reparada", record.ActiveVersion, Progress: 1);
        }
        catch (OperationCanceledException) { State = new(AgerbotModelUpdateStatus.Cancelled, "Reparación cancelada"); }
        catch (Exception exception) { State = new(AgerbotModelUpdateStatus.Failed, exception.Message); }
        finally { EndOperation(); }
    }

    public void Cancel() => _cancellation?.Cancel();

    private async Task InstallAvailableUnsafeAsync(CancellationToken cancellationToken, ulong diskAvailableBytes = ulong.MaxValue)
    {
        var candidate = _availableCandidate ?? throw new InvalidOperationException("No hay una actualización seleccionada.");
        string version = candidate.Manifest.Release.Version;
        State = new(AgerbotModelUpdateStatus.Downloading, $"Descargando Agerbot {version}…", version, candidate.Manifest.Artifact.SizeBytes);
        var progress = new Progress<double>(value =>
            State = new(AgerbotModelUpdateStatus.Downloading, $"Descargando Agerbot {version}… {value:P0}", version, candidate.Manifest.Artifact.SizeBytes, value));
        var downloaded = await _downloads.DownloadAsync(candidate, diskAvailableBytes, progress, cancellationToken);
        State = new(AgerbotModelUpdateStatus.Verifying, "SHA-256 correcto; preparando el candidato", version);
        var installed = await _installation.InstallCandidateAsync(candidate, downloaded.Checkpoint, downloaded.Evaluation, cancellationToken);
        State = new(AgerbotModelUpdateStatus.Verifying, "Probando carga, salud y una generación aislada", version);
        bool valid = await ValidateWithCpuFallbackAsync(installed.CheckpointPath, version, cancellationToken);
        if (!valid)
        {
            await _failedStore.MarkFailedAsync(version, "Falló la validación aislada de carga, salud o generación.", cancellationToken);
            throw new InvalidOperationException("El modelo candidato no superó la prueba aislada del runtime.");
        }
        State = new(AgerbotModelUpdateStatus.WaitingForConversation, "Esperando a que termine la generación activa", version);
        var previous = await CurrentRecordAsync(cancellationToken);
        var record = new AgerbotCurrentModelRecord
        {
            ActiveVersion = version,
            PreviousVersion = previous?.ActiveVersion,
            ActivatedAt = DateTimeOffset.UtcNow,
            CheckpointPath = installed.CheckpointPath,
            ManifestPath = installed.ManifestPath
        };
        State = new(AgerbotModelUpdateStatus.Activating, $"Activando Agerbot {version}", version);
        bool activated = await _activation.ActivateAsync(record, previous, _runtimeActivator, cancellationToken);
        if (!activated) throw new InvalidOperationException("La activación falló; MISIL restauró el modelo anterior.");
        _installation.PruneOldModels(record.ActiveVersion, record.PreviousVersion);
        _availableCandidate = null;
        OnPropertyChanged(nameof(HasAvailableUpdate));
        State = new(AgerbotModelUpdateStatus.Installed, $"Agerbot {version} está instalado y activo", version, Progress: 1);
    }

    private async Task<bool> ValidateWithCpuFallbackAsync(
        string checkpointPath,
        string version,
        CancellationToken cancellationToken)
    {
        bool valid = await _validator.ValidateAsync(
            _settings.Settings.RuntimeExecutablePath,
            checkpointPath,
            version,
            _settings.Settings.RequestedDevice,
            cancellationToken);
        if (!valid
            && !_settings.Settings.RequestedDevice.Equals("cpu", StringComparison.OrdinalIgnoreCase)
            && !string.IsNullOrWhiteSpace(_settings.Settings.CpuRuntimeExecutablePath)
            && !_settings.Settings.CpuRuntimeExecutablePath.Equals(_settings.Settings.RuntimeExecutablePath, StringComparison.OrdinalIgnoreCase))
        {
            valid = await _validator.ValidateAsync(
                _settings.Settings.CpuRuntimeExecutablePath,
                checkpointPath,
                version,
                "cpu",
                cancellationToken);
        }
        return valid;
    }

    private async Task<AgerbotCurrentModelRecord?> CurrentRecordAsync(CancellationToken cancellationToken)
    {
        var stored = await _currentStore.LoadAsync(cancellationToken);
        if (stored != null) return stored;
        if (string.IsNullOrWhiteSpace(_settings.Settings.ActiveModelVersion)
            || string.IsNullOrWhiteSpace(_settings.Settings.CheckpointPath)) return null;
        return new AgerbotCurrentModelRecord
        {
            ActiveVersion = _settings.Settings.ActiveModelVersion,
            PreviousVersion = _settings.Settings.PreviousModelVersion,
            ActivatedAt = DateTimeOffset.UtcNow,
            CheckpointPath = _settings.Settings.CheckpointPath,
            ManifestPath = Path.Combine(Path.GetDirectoryName(_settings.Settings.CheckpointPath)!, "manifest.json")
        };
    }

    private void EndOperation()
    {
        _cancellation?.Dispose();
        _cancellation = null;
        _operation.Release();
    }

    private void OnPropertyChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    public void Dispose()
    {
        Cancel();
        _releases.Dispose();
        _operation.Dispose();
    }
}

File: macos/MISILNative/Sources/MISILNative/AgerbotModelUpdateController.swift
import Foundation

@MainActor
final class AgerbotModelUpdateController: ObservableObject {
    @Published private(set) var phase: AgerbotModelUpdatePhase = .idle("Sin comprobar")
    @Published private(set) var availableVersion: String?
    @Published private(set) var availableSizeBytes: UInt64?
    @Published private(set) var shaVerified = false

    private let releaseService: AgerbotReleaseService
    private let downloadService: AgerbotModelDownloadService
    private let stager: AgerbotArtifactStager
    private let validator: AgerbotCandidateValidator
    private let settingsStore: AgerbotSettingsStore
    private let installationManager: AgerbotInstallationManager
    private let currentStore: AgerbotCurrentModelStore
    private let failedStore: AgerbotFailedVersionStore
    private let activationEngine: AgerbotAtomicActivationEngine
    private let runtimeActivator: AgerbotManagedRuntimeActivator
    private var remoteCandidate: AgerbotRemoteModelCandidate?
    private var operationTask: Task<Void, Never>?

    init(
        releaseService: AgerbotReleaseService = AgerbotReleaseService(),
        downloadService: AgerbotModelDownloadService,
        stager: AgerbotArtifactStager = AgerbotArtifactStager(),
        validator: AgerbotCandidateValidator,
        settingsStore: AgerbotSettingsStore,
        installationManager: AgerbotInstallationManager,
        currentStore: AgerbotCurrentModelStore,
        failedStore: AgerbotFailedVersionStore,
        activationEngine: AgerbotAtomicActivationEngine,
        runtimeActivator: AgerbotManagedRuntimeActivator
    ) {
        self.releaseService = releaseService
        self.downloadService = downloadService
        self.stager = stager
        self.validator = validator
        self.settingsStore = settingsStore
        self.installationManager = installationManager
        self.currentStore = currentStore
        self.failedStore = failedStore
        self.activationEngine = activationEngine
        self.runtimeActivator = runtimeActivator
    }

    var progress: Double { downloadService.progress }
    var isDownloading: Bool { downloadService.isDownloading }

    func scheduleAutomaticCheck() {
        guard settingsStore.settings.automaticModelUpdates else { return }
        startCheck(force: false)
    }

    func startCheck(force: Bool = true) {
        guard operationTask == nil else { return }
        operationTask = Task { [weak self] in
            guard let self else { return }
            await self.checkForUpdates(force: force)
            self.operationTask = nil
        }
    }

    func cancelDownload() {
        downloadService.cancel()
        operationTask?.cancel()
        operationTask = nil
        phase = .cancelled
    }

    func installAvailableUpdate() {
        guard let candidate = remoteCandidate, operationTask == nil else { return }
        operationTask = Task { [weak self] in
            guard let self else { return }
            await self.downloadVerifyAndActivate(candidate)
            self.operationTask = nil
        }
    }

    func rollbackToPrevious() {
        guard let previousVersion = settingsStore.settings.previousModelVersion,
              operationTask == nil else { return }
        operationTask = Task { [weak self] in
            guard let self else { return }
            self.phase = .rollingBack(previousVersion)
            if let candidate = self.installationManager.availableModels.first(where: {
                $0.manifest.model.version == previousVersion
            }) {
                let current = await self.currentRecordFromSettings()
                let target = AgerbotCurrentModelRecord(
                    activeVersion: previousVersion,
                    previousVersion: current?.activeVersion,
                    activatedAt: Date(),
                    checkpointPath: candidate.checkpointURL.path,
                    manifestPath: candidate.manifestURL.path
                )
                let success = await self.activationEngine.activate(
                    candidate: target,
                    previous: current,
                    runtime: self.runtimeActivator
                )
                self.phase = success ? .installed(previousVersion) : .failed("No se pudo volver al modelo anterior.")
            } else {
                self.phase = .failed("La versión anterior ya no está disponible localmente.")
            }
            self.operationTask = nil
        }
    }

    private func checkForUpdates(force: Bool) async {
        if !force, let last = settingsStore.settings.lastUpdateCheckAt,
           Date().timeIntervalSince(last) < 6 * 60 * 60 {
            phase = .idle("Comprobación reciente")
            return
        }
        phase = .checking
        settingsStore.settings.lastUpdateCheckAt = Date()
        do {
            let candidate = try await releaseService.latestCompatible(
                installedVersion: settingsStore.settings.activeModelVersion,
                allowPrerelease: settingsStore.settings.allowPrereleaseModels
            )
            let version = candidate.manifest.release.version
            if await failedStore.contains(version) {
                phase = .failed("Agerbot \(version) falló anteriormente y no se reinstalará automáticamente.")
                return
            }
            remoteCandidate = candidate
            availableVersion = version
            availableSizeBytes = candidate.manifest.artifact.sizeBytes
            phase = .available(version: version, sizeBytes: candidate.manifest.artifact.sizeBytes)
            if settingsStore.settings.automaticModelUpdates {
                await downloadVerifyAndActivate(candidate)
            }
        } catch AgerbotReleaseError.noPublishedRelease {
            phase = .idle("No hay una release pública de Agerbot")
        } catch AgerbotReleaseError.notNewer {
            phase = .idle("Agerbot está actualizado")
        } catch {
            phase = .failed(error.localizedDescription)
        }
    }

    private func downloadVerifyAndActivate(_ candidate: AgerbotRemoteModelCandidate) async {
        let version = candidate.manifest.release.version
        let downloads = installationManager.managedRoot.appendingPathComponent("downloads", isDirectory: true)
        let staging = downloads.appendingPathComponent("staging", isDirectory: true)
        phase = .downloading(version: version, progress: 0)
        let progressTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                self.phase = .downloading(version: version, progress: self.downloadService.progress)
                try? await Task.sleep(for: .milliseconds(150))
            }
        }
        do {
            let downloaded = try await downloadService.download(
                asset: candidate.checkpointAsset,
                to: downloads
            )
            progressTask.cancel()
            phase = .verifying("Comprobando tamaño y SHA-256")
            let stagedModel = try await stager.verifyAndStage(
                downloadedFile: downloaded,
                candidate: candidate,
                stagingRoot: staging
            )
            shaVerified = true
            let installed = try await prepareInstallation(candidate: candidate, stagedModel: stagedModel)
            phase = .verifying("Cargando el modelo candidato en un proceso aislado")
            let valid = await validator.validate(
                checkpointURL: installed.checkpointURL,
                expectedVersion: version,
                projectPath: settingsStore.settings.projectPath
            )
            guard valid else {
                try? await failedStore.markFailed(version: version, reason: "Falló la validación de carga, salud o generación.")
                phase = .failed("El modelo candidato no superó la prueba de runtime.")
                return
            }
            phase = .waitingForConversation
            let previous = await currentRecordFromSettings()
            let record = AgerbotCurrentModelRecord(
                activeVersion: version,
                previousVersion: previous?.activeVersion,
                activatedAt: Date(),
                checkpointPath: installed.checkpointURL.path,
                manifestPath: installed.manifestURL.path
            )
            phase = .activating(version)
            let success = await activationEngine.activate(
                candidate: record,
                previous: previous,
                runtime: runtimeActivator
            )
            if success {
                phase = .installed(version)
                _ = await installationManager.discover(using: settingsStore.settings)
            } else {
                phase = .rollingBack(previous?.activeVersion ?? "modelo anterior")
                phase = .failed("La activación falló; MISIL restauró el modelo anterior.")
            }
        } catch is CancellationError {
            progressTask.cancel()
            phase = .cancelled
        } catch {
            progressTask.cancel()
            phase = .failed(error.localizedDescription)
        }
    }

    private func prepareInstallation(
        candidate: AgerbotRemoteModelCandidate,
        stagedModel: URL
    ) async throws -> AgerbotModelCandidate {
        let version = candidate.manifest.release.version
        let modelsRoot = installationManager.managedRoot.appendingPathComponent("models", isDirectory: true)
        let finalDirectory = modelsRoot.appendingPathComponent(version, isDirectory: true)
        let checkpointURL = finalDirectory.appendingPathComponent("model.pt")
        let manifestURL = finalDirectory.appendingPathComponent("manifest.json")
        if !FileManager.default.fileExists(atPath: checkpointURL.path) {
            try FileManager.default.createDirectory(at: modelsRoot, withIntermediateDirectories: true)
            let temporary = modelsRoot.appendingPathComponent(".candidate-\(UUID().uuidString)", isDirectory: true)
            do {
                try FileManager.default.createDirectory(at: temporary, withIntermediateDirectories: true)
                try FileManager.default.copyItem(at: stagedModel, to: temporary.appendingPathComponent("model.pt"))
                let localManifest = AgerbotModelManifest(
                    schemaVersion: 2,
                    channel: candidate.manifest.channel,
                    model: .init(
                        name: candidate.manifest.model.name,
                        version: version,
                        trainingName: candidate.manifest.model.trainingName,
                        architecture: candidate.manifest.model.architecture,
                        tokenizer: candidate.manifest.model.tokenizer,
                        parameters: candidate.manifest.model.parameters,
                        contextLength: candidate.manifest.model.contextLength
                    ),
                    runtime: .init(
                        minimumVersion: candidate.manifest.runtime.minimumVersion,
                        maximumVersion: candidate.manifest.runtime.maximumVersion
                    ),
                    checkpoint: .init(
                        filename: "model.pt",
                        sizeBytes: candidate.manifest.artifact.sizeBytes,
                        sha256: candidate.manifest.artifact.sha256
                    ),
                    compatibility: .init(
                        devices: candidate.manifest.compatibility.devices,
                        platforms: candidate.manifest.compatibility.platforms
                    ),
                    publishedAt: candidate.manifest.release.publishedAt
                )
                let encoder = JSONEncoder()
                encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
                try encoder.encode(localManifest).write(
                    to: temporary.appendingPathComponent("manifest.json"),
                    options: .atomic
                )
                let evaluationDownload = try await downloadService.download(
                    asset: candidate.evaluationAsset,
                    to: installationManager.managedRoot.appendingPathComponent("downloads", isDirectory: true)
                )
                try FileManager.default.copyItem(
                    at: evaluationDownload,
                    to: temporary.appendingPathComponent("evaluation.json")
                )
                try FileManager.default.moveItem(at: temporary, to: finalDirectory)
            } catch {
                try? FileManager.default.removeItem(at: temporary)
                throw error
            }
        }
        return AgerbotModelCandidate(
            manifest: AgerbotModelManifest(
                schemaVersion: 2,
                channel: candidate.manifest.channel,
                model: .init(
                    name: candidate.manifest.model.name,
                    version: version,
                    trainingName: candidate.manifest.model.trainingName,
                    architecture: candidate.manifest.model.architecture,
                    tokenizer: candidate.manifest.model.tokenizer,
                    parameters: candidate.manifest.model.parameters,
                    contextLength: candidate.manifest.model.contextLength
                ),
                runtime: .init(
                    minimumVersion: candidate.manifest.runtime.minimumVersion,
                    maximumVersion: candidate.manifest.runtime.maximumVersion
                ),
                checkpoint: .init(
                    filename: "model.pt",
                    sizeBytes: candidate.manifest.artifact.sizeBytes,
                    sha256: candidate.manifest.artifact.sha256
                ),
                compatibility: .init(
                    devices: candidate.manifest.compatibility.devices,
                    platforms: candidate.manifest.compatibility.platforms
                ),
                publishedAt: candidate.manifest.release.publishedAt
            ),
            manifestURL: manifestURL,
            checkpointURL: checkpointURL,
            source: .managed
        )
    }

    private func currentRecordFromSettings() async -> AgerbotCurrentModelRecord? {
        if let stored = await currentStore.load() { return stored }
        guard let version = settingsStore.settings.activeModelVersion,
              !settingsStore.settings.checkpointPath.isEmpty else { return nil }
        let checkpoint = URL(fileURLWithPath: settingsStore.settings.checkpointPath)
        return AgerbotCurrentModelRecord(
            activeVersion: version,
            previousVersion: settingsStore.settings.previousModelVersion,
            activatedAt: Date(),
            checkpointPath: checkpoint.path,
            manifestPath: checkpoint.deletingLastPathComponent().appendingPathComponent("manifest.json").path
        )
    }
}
